src/Orleans/Async/Cancellation/GrainCancellationToken.cs

```diff
@@ -0,0 +1,140 @@
+        /// </summary>
+        public CancellationToken CancellationToken => _cancellationTokenSource.Token;
+
+        internal bool IsCancellationRequested => _cancellationTokenSource.IsCancellationRequested;
+
+        internal Task Cancel()
+        {
+            _cancellationTokenSource.Cancel();
+
+            if (Interlocked.Exchange(ref _isInCancellationProcess, 1) != 0)
+            {
+                return TaskDone.Done;
+            }
+
+            if (_targetGrainReferences.IsEmpty)
+            {
+                return TaskDone.Done;
+            }
+
+            var cancellationTasks = new List<Task>();
+
+            foreach (var pair in _targetGrainReferences)
+            {
+                Interlocked.Increment(ref _cancellationRequestsInFlight);
+                var cancellationTask = pair.Value.AsReference<ICancellationSourcesExtension>().CancelTokenSource(this);
+                cancellationTasks.Add(cancellationTask);
+                cancellationTask.ContinueWith(task =>
+                {
+                    if (Interlocked.Decrement(ref _cancellationRequestsInFlight) == 0)
+                    {
+                        Interlocked.Exchange(ref _isInCancellationProcess, 0);
+                    }
+
+                    if (task.IsFaulted) return;
+
+                    // remove reference to which cancellation call has succeded, 
+                    // in order to avoid unnecessary remote call in case of retrying
+                    GrainReference grainRef;
+                    _targetGrainReferences.TryRemove(pair.Key, out grainRef);
+                });
+            }
+
+            return Task.WhenAll(cancellationTasks);
+        }
+
+        internal void AddGrainReference(GrainReference grainReference)
+        {
+            _targetGrainReferences.AddOrUpdate(grainReference.GrainId, id => grainReference, (id, ignore) => grainReference);
+        }
+
+        public void Dispose()
+        {
+            _cancellationTokenSource.Dispose();
+        }
+
+        #region Serialization
+
+        [SerializerMethod]
+        internal static void SerializeGrainCancellationToken(object obj, BinaryTokenStreamWriter stream, Type expected)
+        {
+            var ctw = (GrainCancellationToken)obj;
+            var canceled = ctw.CancellationToken.IsCancellationRequested;
+            stream.Write(canceled);
+            stream.Write(ctw.Id);
+        }
+
+        [DeserializerMethod]
+        internal static object DeserializeGrainCancellationToken(Type expected, BinaryTokenStreamReader stream)
+        {
+            var cancellationRequested = stream.ReadToken() == SerializationTokenType.True;
+            var tokenId = stream.ReadGuid();
+            return new GrainCancellationToken(tokenId, cancellationRequested);
+        }
+
+        [CopierMethod]
+        internal static object CopyGrainCancellationToken(object obj)
+        {
+            return obj; // CancellationToken is value type
```

Comment:
Oh. and deep copying won't  help to cancel the other tokens, so in fact there will be zero cancel calls to execute, as the reference of the token will be pointing to this token (flow: deep copy -> setting of grain reference -> storing in extension -> passing to grain method)